[quant] Generalizing _calculate_dynamic_qparams in quantized test

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36450 [quant] quantized reflection_pad1d
* **#36449 [quant] Generalizing _calculate_dynamic_qparams in quantized test**

Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

Commit:
[quant] Generalizing _calculate_dynamic_qparams in quantized test

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

raghuramank100 commented on torch/testing/_internal/common_quantized.py:
```diff
@@ -43,20 +43,18 @@ def _calculate_dynamic_qparams(X, dtype, reduce_range=False):
     min_val = X.min()
     max_val = X.max()
-    if min_val == max_val:
+    if min_val == max_val == 0:
```
Dont you want to also check for min_val ==max_val also so that we avoid scale=0?

z-a-f replied:
I don't think we need it -- if `min_val == max_val` it will be caught in the later lines, when it says

```
max_val = max(max_val, 0.0)
min_val = min(min_val, 0.0)
```

Because we already checked if they are zero, we will get a valid range `[min_val, max_val]`, where they will be different, while only one of them might be equal to zero. That way, in case we have a tensor with all the same values in it, it will still be correctly representable, as we will have the range `[0, VALUE]`

raghuramank100 reviewed (commented):
I think we need to address one more case here.

z-a-f commented:
> I think we need to address one more case here.

What is the case we need to address?

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

[ghstack-poisoned]

z-a-f commented:
Reverting because of the `test_qelu` breakage https://app.circleci.com/pipelines/github/pytorch/pytorch/160770/workflows/a3ca7391-3ba4-4558-9929-7ff1ffd4794c/

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)

Commit:
Update on "[quant] Generalizing _calculate_dynamic_qparams in quantized test"


Differential Revision: [D20984966](https://our.internmc.facebook.com/intern/diff/D20984966)
